diffusion/datasets/image.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2022 MosaicML Diffusion authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Streaming Image dataset."""
+
+import logging
+from io import BytesIO
+from typing import Callable, Dict, List, Optional, Sequence, Union
+
+from PIL import Image
+from streaming import Stream, StreamingDataset
+from torch.utils.data import DataLoader
+from torchvision import transforms
+
+log = logging.getLogger(__name__)
+
+# Disable PIL max image size limit
+Image.MAX_IMAGE_PIXELS = None
+
+
+class StreamingImageDataset(StreamingDataset):
+    """Streaming dataset for images.
+
+    Args:
+        streams (Sequence[Stream], optional): One or more Streams to stream/cache samples from.
+            ``StreamingImageCaptionDataset`` uses either ``streams`` or ``remote``/``local``. Default:``None``.
+        remote (str, optional): Remote directory (S3 or local filesystem) where dataset is stored. Default: ``None``.
+        local (str, optional): Local filesystem directory where dataset is cached during operation. Default: ``None``.
+        transform (Callable, optional): The transforms to apply to the image. Default: ``None``.
+        image_key (str): Key associated with the image in the streaming dataset. Default: ``'image'``.
+        image_output_key (optional, str): Optional output key for the image. If none, the value of `image_key` will
+            be used. Default: ``None``.
+        return_all_fields (bool, optional): If ``True``, return all fields in the sample. If ``False``, only return
+            the image. Default: ``False``.
+
+        **streaming_kwargs: Additional arguments to pass in the construction of the StreamingDataloader
+    """
+
+    def __init__(
+        self,
+        streams: Optional[Sequence[Stream]] = None,
+        remote: Optional[str] = None,
+        local: Optional[str] = None,
+        transform: Optional[Callable] = None,
+        image_key: str = 'image',
+        image_output_key: Optional[str] = None,
+        return_all_fields: bool = False,
+        **streaming_kwargs,
+    ) -> None:
+
+        # Set defaults for vision-friendly streaming args.
+        streaming_kwargs.setdefault('shuffle_block_size', 1 << 18)
+        streaming_kwargs.setdefault('shuffle_algo', 'py1s')
+
+        super().__init__(
+            streams=streams,
+            remote=remote,
+            local=local,
+            **streaming_kwargs,
+        )
+
+        self.transform = transform
+        self.image_key = image_key
+        self.image_output_key = image_output_key if image_output_key is not None else image_key
+        self.return_all_fields = return_all_fields
+
+    def __getitem__(self, index):
+        sample = super().__getitem__(index)
+        # Image
+        if not isinstance(sample[self.image_key], Image.Image):
+            img = Image.open(BytesIO(sample[self.image_key]))
+        else:
+            img = sample[self.image_key].copy()
```

Comment:
Copy here was to prevent modification of the data in `sample[image_key]` by subsequent transforms since that gets returned if `return_all_fields=True`. Not sure if there is ever a case where this is an issue, so I can remove it if you think it's unnecessary.